Fix null reference when module is null in hosting scenarios

adityapatwardhan opened the pull request:
# PR Summary
Fixes #9390 

If the `CommandInfo.Module` is null, then default to a new `PSMarkdownOptionInfo`.

## PR Context

In hosting scenarios as module are loaded as snapins, the `CommandInfo.Module` is null. In those cases we default back to default `PSMarkdownOptionInfo`.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

adityapatwardhan committed:
Fix null reference when module is null in hosting scenarios

adityapatwardhan committed:
Add test for hosting scenario

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/MarkdownCmdlets.Tests.ps1:
```diff
@@ -546,4 +546,25 @@ bool function()`n{`n}
+                }) | Out-Null
+
+            $ps.Invoke()
+            $ps.Dispose()
```
Usually we put Dispose in `finally` block of try-catch.

adityapatwardhan replied:
Fixed.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/MarkdownCmdlets.Tests.ps1:
```diff
@@ -546,4 +546,25 @@ bool function()`n{`n}
+                        $output =  '# test' | ConvertFrom-Markdown
+                        $output.Html | Should -BeExactly '<h1 id="test">test</h1>'
+                    } catch {
+                        $_ | Out-Default
```
I wonder why do we create this output if then send it to Out-Null?

adityapatwardhan replied:
Removed

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromMarkdownCommand.cs:
```diff
@@ -66,12 +66,8 @@ public class ConvertFromMarkdownCommand : PSCmdlet
-            {
-                throw new InvalidOperationException();
-            }
+            _mdOption = this.CommandInfo.Module?.SessionState.PSVariable.GetValue("PSMarkdownOptionInfo", new PSMarkdownOptionInfo()) as PSMarkdownOptionInfo;
```
Could you please clarify for my education why is `this.CommandInfo.Module` null in the scenario?

adityapatwardhan replied:
Its part of the PR context.

iSazonov replied:
My question is why this is null if it is running in new runspace. I do not understand the difference and therefore I am afraid that there may be a bug.

SteveL-MSFT replied:
https://docs.microsoft.com/en-us/dotnet/api/system.management.automation.commandinfo.module?view=powershellsdk-1.1.0#System_Management_Automation_CommandInfo_Module indicates that this shouldn't be null since this cmdlet is part of a module.  Seems like there's a bug somewhere else.  I don't think this would block this PR, but we should have an issue open to investigate this.

iSazonov replied:
We already have #9390 for tracking. If it is "a bug somewhere else". Since we are at the beginning of a new milestone we could only add pending test in the PR without temporary fix.

adityapatwardhan replied:
My assumption of it being because of snapins seems to be wrong. @daxian-dbw might have some insights here.

SteveL-MSFT replied:
Yes, if it's a snapin, it won't show up as a module

daxian-dbw replied:
Yep, it's because powershell built-in modules are loaded as snapins with `[runspacefactory]::CreateRunspacePool(1, 2, $Host)`, as well as `[runspacefactory]::CreateRunspace()`.
This issue repros with the following script too:
```
PS:100> $ps = [powershell]::Create()
PS:101> $ps.AddScript("'# test' | convertfrom-markdown").Invoke()
PS:102> $ps.Streams.Error
convertfrom-markdown : Object reference not set to an instance of an object.
At line:1 char:12
+ '# test' | convertfrom-markdown
+            ~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [ConvertFrom-Markdown], NullReferenceException
+ FullyQualifiedErrorId : System.NullReferenceException,Microsoft.PowerShell.Commands.ConvertFromMarkdownCommand
```

iSazonov replied:
Thanks for clarify. If snapins is deprecated should we change loading built-in modules too? If possible this would be the best fix.

daxian-dbw replied:
The public interface of PSSnapin is deprecated, meaning user cannot define a snapin to work with PowerShell, but the snapin itself is still useful, especially in the hosting scenario.
Loading as snapin doesn't require the modules to be present, so for some hosting scenario, this is convenient as the hosting application doesn't need to ship the built-in powershell modules along with the application.

iSazonov replied:
I wonder - we can create an assembly with Add-Type and load it with Import-Module. We use this in our tests without creating module folder and etc.

adityapatwardhan committed:
Address Ilya's comments

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/MarkdownCmdlets.Tests.ps1:
```diff
@@ -549,22 +549,26 @@ bool function()`n{`n}
+                $ps.AddScript( {
+                        try {
+                            $output = '# test' | ConvertFrom-Markdown
+                            $output.Html.trim() | Should -BeExactly '<h1 id="test">test</h1>'
```
Interesting, we run Pester (`Should -BeExactly`) in another thread - is it right way? I am inclined to think that the test is now dummy.

adityapatwardhan replied:
Moved the validation outside the AddScript

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromMarkdownCommand.cs:
```diff
@@ -66,12 +66,8 @@ public class ConvertFromMarkdownCommand : PSCmdlet
-            {
-                throw new InvalidOperationException();
-            }
+            _mdOption = this.CommandInfo.Module?.SessionState.PSVariable.GetValue("PSMarkdownOptionInfo", new PSMarkdownOptionInfo()) as PSMarkdownOptionInfo;
```
I think this is not the right fix.
1. `Get/Set-MarkdownOption` both use the same code to set and get the option variable.
2. Need to make sure `Get/Set-MarkdownOption` continue to work even the utility.dll is loaded as snapin.

adityapatwardhan replied:
Updated to use a concurrentdictionary

adityapatwardhan committed:
Update the cmdlets to store the options per runspace

* Instead of using the module scope variable, we now use a dictionary with the runspace as the key to store the options.
* Update tests accordingly.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/MarkdownCmdlets.Tests.ps1:
```diff
@@ -546,4 +516,98 @@ bool function()`n{`n}
+
+                $options = $ps.Invoke()
+
+                $options.Header1 | Should -BeExactly "[7m"
```
I'd put on first place to get more clear message if `Get-MarkdownOption` fail:
```suggestion
                $options | Should -Not -Be $null
```

adityapatwardhan replied:
Fixed.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/MarkdownCmdlets.Tests.ps1:
```diff
@@ -546,4 +516,98 @@ bool function()`n{`n}
+
+                $options = $ps.Invoke()
+
+                $options.Header1 | Should -BeExactly "[93m"
```
The same.

adityapatwardhan replied:
Fixed

adityapatwardhan commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MarkdownOptionCommands.cs:
```diff
@@ -256,7 +260,38 @@ public class GetMarkdownOptionCommand : PSCmdlet
+
+        static PSMarkdownOptionInfoCache()
+        {
+            markdownOptionInfoCache = new ConcurrentDictionary<string, PSMarkdownOptionInfo>();
```
Fixed.

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Utility/MarkdownCmdlets.Tests.ps1:
```diff
@@ -546,4 +516,98 @@ bool function()`n{`n}
+                            $output = '# test' | ConvertFrom-Markdown
+                            $output.Html.trim()
+                        } catch {
+                            throw $_
```
Removed.

adityapatwardhan committed:
Address Ilya's comments

adityapatwardhan commented:
@daxian-dbw Can you have a look again.

adityapatwardhan committed:
Use module scope variable when moduleInfo is available

When moduleInfo is available we use the module scope variable.
But, when it is not available, we use a concurrent dictionary with runspace instance ID as the key to store the options.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MarkdownOptionCommands.cs:
```diff
@@ -173,11 +175,23 @@ protected override void EndProcessing()
+                this.CommandInfo.Module.SessionState.PSVariable.Set(MarkdownOptionInfoVariableName, mdOptionInfo);
+                setOption = mdOptionInfo;
+            }
+            else // If we don't have a moduleInfo, like in PowerShell hosting scenarios, use a concurrent dictionary.
```
I guess CodeFactor will reports a style issue for the comment.
We could move it before line 188.

daxian-dbw replied:
As mentioned above, comments need to be make more clear.

adityapatwardhan replied:
Refactored this.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MarkdownOptionCommands.cs:
```diff
@@ -256,7 +270,43 @@ public class GetMarkdownOptionCommand : PSCmdlet
+            var option = this.CommandInfo.Module == null
+                // If we don't have a moduleInfo, like in PowerShell hosting scenarios, use a concurrent dictionary.
+                ? PSMarkdownOptionInfoCache.Get(Runspace.DefaultRunspace.InstanceId)
+                // If we have the moduleInfo then store are module scope variable
```
I guess CodeFactor will reports style issues for the comments.
We could move both to line  273.

adityapatwardhan replied:
Fixed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MarkdownOptionCommands.cs:
```diff
@@ -256,7 +270,43 @@ public class GetMarkdownOptionCommand : PSCmdlet
+            {
+                // no option cache so cache and return the default PSMarkdownOptionInfo
+                var newOptionInfo = new PSMarkdownOptionInfo();
+                return markdownOptionInfoCache.GetOrAdd(runspaceId, (key) => newOptionInfo);
```
It seems we have no need to do `TryGetValue` before.
```c#
     internal static PSMarkdownOptionInfo Get(Guid runspaceId)
    {
         return markdownOptionInfoCache.GetOrAdd(runspaceId, (key) => new PSMarkdownOptionInfo());
    }
```

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromMarkdownCommand.cs:
```diff
@@ -66,11 +66,16 @@ public class ConvertFromMarkdownCommand : PSCmdlet
+            // If we have the moduleInfo then store are module scope variable
+            if (this.CommandInfo.Module != null)
+            {
+                _mdOption = this.CommandInfo.Module?.SessionState.PSVariable.GetValue("PSMarkdownOptionInfo", new PSMarkdownOptionInfo()) as PSMarkdownOptionInfo;
```
You have checked `ComamndInfo.Module != null`, so no need to use null-conditional operator here.

adityapatwardhan replied:
Fixed with refactor

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromMarkdownCommand.cs:
```diff
@@ -66,11 +66,16 @@ public class ConvertFromMarkdownCommand : PSCmdlet
+                _mdOption = this.CommandInfo.Module?.SessionState.PSVariable.GetValue("PSMarkdownOptionInfo", new PSMarkdownOptionInfo()) as PSMarkdownOptionInfo;
+                _mdOption = _mdOption ?? new PSMarkdownOptionInfo();
+            }
+            else // If we don't have a moduleInfo, like in PowerShell hosting scenarios, use a concurrent dictionary.
```
> like in PowerShell hosting scenarios,

The comment needs to be more clear. Not all hosting scenarios will have this problem. Only those where built-in assemblies are loaded as snapins.

adityapatwardhan replied:
Updated.

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MarkdownOptionCommands.cs:
```diff
@@ -256,7 +270,43 @@ public class GetMarkdownOptionCommand : PSCmdlet
         {
-            WriteObject(this.CommandInfo.Module.SessionState.PSVariable.GetValue(MarkdownOptionInfoVariableName, new PSMarkdownOptionInfo()));
+            var option = this.CommandInfo.Module == null
+                // If we don't have a moduleInfo, like in PowerShell hosting scenarios, use a concurrent dictionary.
```
Same here, comment needs to be made more clear

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ConvertFromMarkdownCommand.cs:
```diff
@@ -66,11 +66,16 @@ public class ConvertFromMarkdownCommand : PSCmdlet
-            _mdOption = this.CommandInfo.Module.SessionState.PSVariable.GetValue("PSMarkdownOptionInfo", new PSMarkdownOptionInfo()) as PSMarkdownOptionInfo;
-
-            if (_mdOption == null)
+            // If we have the moduleInfo then store are module scope variable
```
I think we need comments that can better explain why we have different ways to handle this.

adityapatwardhan replied:
Fixed.

daxian-dbw commented:
Also talked to @adityapatwardhan offline, and we agreed it's better to move all `Get/Set` `psmarkdownoption` code to a single class.

adityapatwardhan committed:
Move get set code for options to internal class

adityapatwardhan committed:
Update comments to give more detail

adityapatwardhan commented:
@daxian-dbw I have updated to address your comments. Please have another look.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MarkdownOptionCommands.cs:
```diff
@@ -270,28 +257,30 @@ public class GetMarkdownOptionCommand : PSCmdlet
+            // If we have the moduleInfo then store are module scope variable
+            if (command.Module != null)
+            {
+                return command.Module.SessionState.PSVariable.GetValue(MarkdownOptionInfoVariableName, new PSMarkdownOptionInfo()) as PSMarkdownOptionInfo;
```
I wonder that the API to force creating default value before trying to get a value.

adityapatwardhan commented:
@daxian-dbw Ping..

daxian-dbw commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/MarkdownOptionCommands.cs:
```diff
@@ -256,7 +257,61 @@ public class GetMarkdownOptionCommand : PSCmdlet
+            {
+                // no option cache so cache and return the default PSMarkdownOptionInfo
+                var newOptionInfo = new PSMarkdownOptionInfo();
+                return markdownOptionInfoCache.GetOrAdd(Runspace.DefaultRunspace.InstanceId, (key) => newOptionInfo);
```
You can use the overload `GetOrAdd(TKey, TValue)`, so no need to create a delegate.

adityapatwardhan replied:
Fixed.

adityapatwardhan committed:
Address Dongbo's comments

adityapatwardhan commented:
@daxian-dbw This is ready for merging.
